### PR TITLE
Remove getdents64() API, implement opendir()/readdir()

### PR DESCRIPTION
**Pull request by rianhunter**
This is related to issue #7649 (specifically point 7). The `getdents64()` API is not easily portable across different platforms due to the fact that there is no standardized way to read dirents from a raw file descriptor or seek a raw file descriptor to a specific dirent, and in practice each major platform does it differently with different contraints. Platforms tend to only support an `opendir()`/`readdir()`-based directory traversal API.

To make it easier to run Emscripten-generated WASM files in non-browser contexts, this change makes generated WASM files import `opendir()`/`readdir()` functions from the host environment instead of the Linuxy `getdents64()` syscall. This makes the host functions that support those calls a lot less hacky.

Note that this does not change the dirent ABI, which continues to have the same structure in memory as before. This allows us to continue using higher level musl code that interacts with directories, like `scandir()`.

### CODE REVIEW

**Comment by sbc100** [comment]
This seems like the kind of thing that moving the low level interface to WASI might help with.

I'm slightly hesitant to move stuff out of C code and into JS libraries, and I'm not sure how this general direction aligns with emscriptens long term goals.  What kind of non-web contexts are you considering and what kind of support from emscripten will you be expecting for thse non-web-contexts?  Presumably we would need to start testing in these contexts if we want to avoid breaking you?

**Comment by rianhunter** [comment]
I share your hesitation with moving code out to JS but pushing the API layer to opendir/readdir is more portable than the syscall layer (since it is spec'd by posix, while the syscall isn't, and in practice isn't portable either).

I plan to publish a spec similar to WASI with the explicit intent of being a fully conforming POSIX implementation. WASI has similar but conflicting goals in that regard. I want that spec to evolve with Emscripten which is why I am making changes piece by piece to make sure it works well in practice before publishing a spec without an implementation.

File: src/library_syscall.js [hunk 1]
```diff
@@ -1059,34 +1059,7 @@ var SyscallsLibrary = {
...
-      }
-      {{{ makeSetValue('dirp + pos', C_STRUCTS.dirent.d_ino, 'id', 'i32') }}};
-      {{{ makeSetValue('dirp + pos', C_STRUCTS.dirent.d_off, 'stream.position', 'i32') }}};
-      {{{ makeSetValue('dirp + pos', C_STRUCTS.dirent.d_reclen, C_STRUCTS.dirent.__size__, 'i16') }}};
-      {{{ makeSetValue('dirp + pos', C_STRUCTS.dirent.d_type, 'type', 'i8') }}};
-      stringToUTF8(name, dirp + pos + {{{ C_STRUCTS.dirent.d_name }}}, 256);
-      pos += {{{ C_STRUCTS.dirent.__size__ }}};
-    }
-    return pos;
+    return -ERRNO_CODES.ENOSYS; // unsupported feature, use readdir()
```
Comment (sbc100) [inline src/library_syscall.js]: Can we simply remove this completely?

File: tools/system_libs.py [hunk 2]
```diff
@@ -232,6 +232,12 @@ def create_libc(libname):
         'faccessat.c',
     ]
 
+    # any dirent file that digs into DIR structure
```
Comment (sbc100) [inline tools/system_libs.py]: Can you say why in the comment.  Something like: ".. directory handling in implemented JS"?

File: src/library.js [hunk 3]
```diff
@@ -51,6 +51,221 @@ LibraryManager.library = {
...
+      stream.getdentsposition = 0;
+      stream.inc_pos_before_read = 0;
+
+      return stream.fd + 1;
+    } catch (e) {
+      FS.handleFSError(e);
+      return 0;
+    }
+  },
+
```
Comment (sbc100) [inline src/library.js]: Any reason all this can't live in library_fs.js?

**Comment by sbc100** [comment]
OK I do see you point.

File: src/library_syscall.js (hunk 1 above)
Comment (rianhunter) [reply src/library_syscall.js]: Yes, we may be able to do that. I left it in there for future code spelunkers who might wonder why every other syscall from musl is defined in this file except for this one. I'll just make it a comment

File: tools/system_libs.py (hunk 2 above)
Comment (rianhunter) [reply tools/system_libs.py]: Sure

File: src/library_syscall.js (hunk 1 above)
Comment (sbc100) [reply src/library_syscall.js]: I'm think just remove it.. or perhaps leave a comment in its place?

Are there no other missing syscalls?

File: src/library.js (hunk 3 above)
Comment (rianhunter) [reply src/library.js]: Hmm, I was under the impression that `library_fs.js` was an Emscripten-internal support library while `library.js` was the C library. I could move this functionality into `library_fs.js` but then I would have to duplicate some support code in other FS implementations (e.g. asmfs).

Comment (sbc100) [reply src/library.js]: I'm not an expert in this area.  I'll defer to @kripken

**Comment by kripken** [comment]
I'd like to understand more the direction this is heading. In general, moving code from musl C to custom JS is not something we prefer to do, because musl's code is very reliable and because it's more efficient (small/fast) in C.

It seems the idea is to make the ABI boundary the libc API, instead of a lower-level syscall API?

**Comment by rianhunter** [comment]
> It seems the idea is to make the ABI boundary the libc API, instead of a lower-level syscall API?

For the purposes of this specific change, it's only to move the ABI boundary from the syscall level to libc because the getdents{64}() syscall isn't generally portable across different POSIX system. I could go into the specific reasons why if you want but for the sake of brevity, assume that is true.

Long term my goal isn't to replace every ABI boundary with libc calls, just the ones where there is a strong impedance mismatch with underlying implementations like `readdir()`/`getdents()` and the ABI change doesn't require disruptive/drastic changes to the Emscripten codebase. As far as a written spec goes, my goal is to document the ABI boundaries that Emscripten currently implements, not to change Emscripten to conform to a yet unwritten spec.

This doesn't preclude a potential future effort to reconcile Emscripten's ABI with WASI, when/if that happens, whatever document I produce will reflect that ABI change. The purpose of the document is to make implementing POSIX conformance for Emscripten-generated binaries clear to third-parties.

In terms of abstract purity, it would be nice to make the ABI more directly reflect POSIX/libc for all the lower-level calls, but that isn't an explicit goal. That is largely a cosmetic thing and not my concern, though if that is what is desired for Emscripten's internal interface, it can be done.

The only other change on my radar from here is to make `off_t` and `ino_t` 64-bits, and to maybe resurrect #7799 (though something tells me it was abandoned for a reason)

**Comment by sbc100** [comment]
I agree with the general direction of this change since getdents() is a very linux-specific thing.

@kripken what do you think about putting these new JS functions in library.js vs library_fs.js?    Perhaps I'm missing the division of responsibility between those two files but it feels like these are certainly fs related.. I'm just not sure if that mean that should for sure live in the _fs.js file?

File: src/library.js (hunk 3 above)
Comment (kripken) [reply src/library.js]: I think the main question is, should this code be present when the user sets `-s FILESYSTEM=0`. If not, then `library_fs.js` is the right place.

(We should probably move some other `library` code into `library_fs`, btw.)

**Comment by kripken** [comment]
Ok - if you two agree that this makes sense to do specifically for `getdents64`, and is not part of a larger move of code from musl C to custom JS, then I'm not opposed to this.

File: emcc.py [hunk 4]
```diff
@@ -1132,13 +1132,19 @@ def check(input_file):
       shared.Settings.FILESYSTEM = 0
       shared.Settings.FETCH = 1
       options.js_libraries.append(shared.path_from_root('src', 'library_asmfs.js'))
+      # library_asmfs.js requires emscripten_asmfs_readdir
+      shared.Settings.EXPORTED_FUNCTIONS += ['_emscripten_asmfs_readdir']
+      shared.Settings.EXPORTED_FUNCTIONS += ['___errno_location']
```
Comment (kripken) [inline emcc.py]: the changes in this file seem unrelated to the PR?

**Comment by kripken** [comment]
The asmfs changes seem separable from the main part of this PR, or am I missing something?

We need to verify that we have good tests, or add some, for the functionality we are moving to custom JS code.

File: emcc.py (hunk 4 above)
Comment (rianhunter) [reply emcc.py]: These changes are necessary to support the changes in `library_asmfs.js` related to `readdir()`. The changes below this aren't specifically related to this PR but necessary for the tests to pass. I yoinked them from #8454

**Comment by rianhunter** [comment]
@kripken are the browser tests corrects functional, or is my change breaking them? The logs aren't clear, and those tests succeed on my machine.

There are already a good amount of tests that test `readdir()` and `opendir()`, I'm not sure how I can add something that would test these specific changes in a more thorough way. Any suggestions?

**Comment by sbc100** [comment]
> @kripken are the browser tests corrects functional, or is my change breaking them? The logs aren't clear, and those tests succeed on my machine.
> 
> There are already a good amount of tests that test `readdir()` and `opendir()`, I'm not sure how I can add something that would test these specific changes in a more thorough way. Any suggestions?

I don't think this should require new tests, since its mostly and internal refactor and as you say those functions should already be tested.

File: emcc.py [hunk 5]
```diff
@@ -1132,13 +1132,19 @@ def check(input_file):
...
 
     if shared.Settings.FETCH and final_suffix in JS_CONTAINING_ENDINGS:
       input_files.append((next_arg_index, shared.path_from_root('system', 'lib', 'fetch', 'emscripten_fetch.cpp')))
       next_arg_index += 1
       options.js_libraries.append(shared.path_from_root('src', 'library_fetch.js'))
       if shared.Settings.USE_PTHREADS:
         shared.Settings.FETCH_WORKER_FILE = unsuffixed(os.path.basename(target)) + '.fetch.js'
+        # This is a hack for make_fetch_worker(), which requires
+        # emscripten_is_main_runtime_thread, which is called from emscripten_fetch()
+        shared.Settings.EXPORTED_FUNCTIONS += ['_emscripten_fetch']
```
Comment (sbc100) [inline emcc.py]: I don't understand why this change would be needed to make the tests pass for this change.  Can you explain a bit more how this is related?

File: system/lib/fetch/asmfs.cpp [hunk 6]
```diff
@@ -1879,78 +1879,61 @@ long __syscall197(int which, ...) // SYS_fstat64
 // TODO: syscall207: fchown32
 // TODO: syscall212: chown32
 
-long __syscall220(int which, ...) // getdents64 (get directory entries 64-bit)
```
Comment (sbc100) [inline system/lib/fetch/asmfs.cpp]: Can you use that same trick and just override readdir here?   Thus avoiding the extra dependencies and JS hop?

Before this change the native version of `__syscall220` would end up meaning it was never imported.. after this change the native readdir here would mean `readdir` was never imported?  Would that work?

File: src/library.js (hunk 3 above)
Comment (sbc100) [reply src/library.js]: I don't think these function should exist for users who set `-s FILESYSTEM=0`, so yes it sounds like they should be in the `_fs.js` file

File: emcc.py (hunk 5 above)
Comment (rianhunter) [reply emcc.py]: Sure, it was non-obvious to me for a few hours as well :) In the ASMFS/USE_PTHREADS/test_readdir.c test in test_browser.py, the only FS code that is used is opendir() and friends. This causes `opt` to dead code eliminate `static open()` in asmfs.cpp, and in turn removing the only dependence on `emscripten_fetch()`, in turn removing the only dependence on `emscripten_is_main_runtime_thread()`, which later causes `make_fetch_worker()` fail.

File: system/lib/fetch/asmfs.cpp (hunk 6 above)
Comment (rianhunter) [reply system/lib/fetch/asmfs.cpp]: Hmm not sure I quite understand. What exactly are you suggesting? Are you saying implement opendir()/readdir() in this file directly?

File: emcc.py (hunk 5 above)
Comment (sbc100) [reply emcc.py]: I see.   But if its `emscripten_is_main_runtime_thread` that you need then why not depend on that?

File: system/lib/fetch/asmfs.cpp (hunk 6 above)
Comment (sbc100) [reply system/lib/fetch/asmfs.cpp]: Yes, well its just readdir right?  It looks like the point of this function is to override readdir()?   So why not just override it here, rather than going out to JS and the calling back into this new function called `emscripten_asmfs_readdir` which didn't exist before?

With this override the linker should then choose this version rather than the JS version, no?

Comment (rianhunter) [reply system/lib/fetch/asmfs.cpp]: Yeah you should be right. I originally opted to not go in that direction to reduce code duplication but if I move the opendir/readdir code to library_fs, then there will be code duplication anyway. If I put the main code in library_fs and the user sets FILESYSTEM=0, do you expect that to be a compile-time error or a runtime-error?

Comment (sbc100) [reply system/lib/fetch/asmfs.cpp]: Perhaps I'm misunderstanding something.    My understanding is that this function here is designed to replace the normal "readdir" when the users whats to use asmfs.

In that case why not just have this version be called "readdir" and for those users they will never import a "readdir" from the output because they are effectively providing their own.  How does that make for code duplication?

Sorry for all the feedback, I'll try to understand this code a little better tomorrow :)

**Comment by kripken** [comment]
The `zzz_emrun` failure is a known random error. The other failures look like they are because of the changes in this PR.

File: system/lib/fetch/asmfs.cpp (hunk 6 above)
Comment (rianhunter) [reply system/lib/fetch/asmfs.cpp]: In the ASMFS case, in the current version of the diff, the path from a normal call to readdir() is this:

readdir() -> library.js -> library_asm.js -> asmfs.cpp

This allows me to use all the bookkeeping logic in library.js. What your suggesting would change that to this:

readdir() -> asmfs.cpp

But I would have to reimplement the `DIR*` bookkeeping logic in asmfs.cpp, which is fine, but then we'd have two readdir() implementations, one in C and one in JS. I think what you're suggesting is more in line with how asmfs.cpp was meant to be, so I will change it to that.

File: emcc.py (hunk 5 above)
Comment (rianhunter) [reply emcc.py]: Just because the condition is on `Settings.FETCH` so it seemed more appropriate.

**Comment by rianhunter** [comment]
After talking with @sunfishcode a bit more about things on the WASI side, I'm going to abandon this PR. It's indeed possible to successfully and safely emulate syscall220 on non-Linux systems, as long as the user program is POSIX conforming. The only downside is extra bookkeeping and a few ifdefs in the host implementation. It would be nice to have a simple passthrough to native opendir/readdir but it's not necessary.

What's more important is changing `ino_t` and `off_t` to be 64-bits, so I will be pursuing that instead.
